### ISSUE TITLE
HTTP: Make headerRegex Accept a CharSequence as Header Name

### ISSUE THREAD

**Issue by jGleitz** [gatling/gatling#3969]
Gatling 3.4 deprecates `io.gatling.http.HeaderNames` in favour of `io.netty.handler.codec.http.HttpHeaderNames`. Hence, all header name constants are now `AsciiStrings`s, not `String`s. However, the `headerRegex` check still demands a `String`. It should be changed to `CharSequence` so it can be used with `io.netty.handler.codec.http.HttpHeaderNames` more easily.

In a nutshell, I want to be able to write:

```scala
http("example")
    .get("/my/url")
    .check(headerRegex(LOCATION, ".*"))
```

currently, I am forced to write:

```scala
http("example")
    .get("/my/url")
    .check(headerRegex(s"$LOCATION", ".*"))
```

**Comment by jGleitz** [comment]
Please note that this issue is actually a regression: Since `HeaderNames` now returns `AsciiString`s instead of `String`s, code that compiled with Gatling 3.3 no longer compiles with Gatling 3.3

**Comment by slandelle** [comment]
> compiled with Gatling 3.3 no longer compiles with Gatling 3.3

You probably mean 3.4. That's expected and [documented](https://gatling.io/docs/current/migration_guides/3.3-to-3.4/), not a regression.

**Comment by jGleitz** [comment]
> That's expected

but why? You could have ensured backwards compatibility by making the `HeaderNames` return `String`s, just like you did for `HeaderValues`.

> and documented

it is actually not documented: The document does not mention that the `HeaderNames` changed their types.

**Comment by slandelle** [comment]
Binary compatibility is only guaranteed across fixes, not minor versions. Period.
Note that FrontLine 1.12 now supports multiple Gatling versions (3.3 and 3.4).

**Comment by jGleitz** [comment]
I am sorry if I gave you the impression that I wanted to argue or demand something. I don’t. I just want to report the problems I am having in the hope that we can make Gatling better for everyone.

> Binary compatibility is only guaranteed across fixes, not minor versions. Period.

My point here is that Gatling broke _source code compatibility_ over a minor version. Code that compiled previously does not compile anymore. Now I think it would be helpful if this change was at least documented very clearly.

I see the point that making `HeaderNames` strings again would break binary compatibility. On the other hand, this needs to be weighed against the impact of the source code incompatibility, I think.

Back to my initial ticket: Can you add an overload of `headerRegex` that accepts a `CharSequence` and mark the existing `headerRegex` `@Deprecated`? That would keep binary compatibility and solve the issue.

**Comment by slandelle** [comment]
> My point here is that Gatling broke source code compatibility over a minor version. Code that compiled previously does not compile anymore. Now I think it would be helpful if this change was at least documented very clearly.

Now I see your point, sorry.

So indeed, ``headerRegex`` should accept a ``CharSequence`` and we missed that. Sadly, this is not something we can fix until Gatling 3.5.0 because of binary compatibility.

> Back to my initial ticket: Can you add an overload of headerRegex that accepts a CharSequence and mark the existing headerRegex @Deprecated? That would keep binary compatibility and solve the issue, I think.

Sadly no, that wouldn't work, because `headerRegex` doesn't take a `String` but an `Expression[String]`. We can't add an overload that would take an `Expression[CharSequence]` because of Java's type erasure (generics' type parameters are lost once compiled).

The best we can do for now is to advertise the workaround in the migration guide.

```scala
http("example")
    .get("/my/url")
    .check(headerRegex(LOCATION.toString, ".*"))
```

> You could have ensured backwards compatibility by making the HeaderNames return Strings, just like you did for HeaderValues.

No. The code is to be able to benefit from from [AsciiString's cached hashcode](https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/AsciiString.java#L1364) when doing by name lookups in [DefaultHttpHeaders](https://github.com/netty/netty/blob/4.1/codec-http/src/main/java/io/netty/handler/codec/http/DefaultHttpHeaders.java).

We don't have the same performance concern when dealing with header values. The only goal when deprecating HeaderValues was to avoid duplicating Netty's constants.
Now I'm considering reverting the deprecation.

BTW, based on the example you provided, I suspect you're disabling redirect in order to capture some data in the Location header. You shouldn't do that. Instead, you should let Gatling follow the redirect and use the `currentLocationRegex` check to capture data from the landing url.

**Comment by jGleitz** [comment]
Thank you for all the explanations! Given the circumstance, I think documenting the change clearly and fixing `headerRegex` in 3.5.0 is a sensible solution.

Thank you for addressing the issue!

---

Regarding the `Location` header and disabling redirects: Thank you for the pointer! We are aware of `currentLocationRegex` and use it in other load tests. For this specific project, we decided to disable redirects because we sometimes only want to check that the redirect was sent, but we don’t want to actually follow it. To have consistent behaviour, we disabled redirects globally.

Is there any downside to doing the redirects manually apart from it being more involved than just having Gatling do it?
